clustered/server/src/main/java/org/ehcache/clustered/server/store/ClusterTierActiveEntity.java

```diff
@@ -586,40 +594,74 @@ public void synchronizeKeyToPassive(PassiveSynchronizationChannel<EhcacheEntityM
     if (concurrencyKey == DEFAULT_KEY) {
       stateService.getStateRepositoryManager().syncMessageFor(storeIdentifier).forEach(syncChannel::synchronizeToPassive);
     } else {
+      BlockingQueue<EhcacheDataSyncMessage> messageQ = new SynchronousQueue<>();
       int segmentId = concurrencyKey - DEFAULT_KEY - 1;
-      Long dataSizeThreshold = Long.getLong(SYNC_DATA_SIZE_PROP, DEFAULT_SYNC_DATA_SIZE_THRESHOLD);
-      AtomicLong size = new AtomicLong(0);
-      ServerSideServerStore store = stateService.getStore(storeIdentifier);
-      final AtomicReference<Map<Long, Chain>> mappingsToSend = new AtomicReference<>(new HashMap<>());
-      store.getSegmentKeySets().get(segmentId)
-        .forEach(key -> {
-          final Chain chain;
-          try {
-            chain = store.get(key);
-          } catch (TimeoutException e) {
-            throw new AssertionError("Server side store is not expected to throw timeout exception");
-          }
-          for (Element element : chain) {
-            size.addAndGet(element.getPayload().remaining());
-          }
-          mappingsToSend.get().put(key, chain);
-          if (size.get() > dataSizeThreshold) {
-            syncChannel.synchronizeToPassive(new EhcacheDataSyncMessage(mappingsToSend.get()));
-            mappingsToSend.set(new HashMap<>());
-            size.set(0);
+      CompletableFuture<Void> asyncGet = CompletableFuture.runAsync(() -> doGetsForSync(segmentId, messageQ));
+      try {
+        boolean done = false;
+        while (!done) {
+          EhcacheDataSyncMessage msg = messageQ.take();
+          if (msg instanceof LastDataSyncMessage) {
+            done = true;
+          } else {
+            syncChannel.synchronizeToPassive(msg);
           }
-        });
-      if (!mappingsToSend.get().isEmpty()) {
-        syncChannel.synchronizeToPassive(new EhcacheDataSyncMessage(mappingsToSend.get()));
-        mappingsToSend.set(new HashMap<>());
-        size.set(0);
+        }
+        asyncGet.get();
+      } catch (InterruptedException | ExecutionException e) {
+        throw new RuntimeException(e);
       }
     }
     sendMessageTrackerReplication(syncChannel, concurrencyKey - 1);
 
     LOGGER.info("Sync complete for concurrency key {}.", concurrencyKey);
   }
 
+  private void doGetsForSync(int segmentId, BlockingQueue<EhcacheDataSyncMessage> messageQ) {
+    if (dataSizeThreshold == null) {
+      // do it only once
+      dataSizeThreshold = Long.getLong(SYNC_DATA_SIZE_PROP, DEFAULT_SYNC_DATA_SIZE_THRESHOLD);
+      dataGetsThreshold = Integer.getInteger(SYNC_DATA_GETS_PROP, DEFAULT_SYNC_DATA_GETS_THRESHOLD);
+    }
+    ServerSideServerStore store = stateService.getStore(storeIdentifier);
+    int numGets = 0;
+    long size = 0;
+    Map<Long, Chain> mappingsToSend = new HashMap<>();
```

Comment:
Should the map be sized to `dataGetsThreshold`?

Reply:
that would cause unnecessary memory usage for very small caches, won't it?

Reply:
Not really is the number of mappings to send is always `numGets`. Or size is generally reached first?

Reply:
Fair enough then

Reply:
@henri-tremblay Actually you have a point. I did some digging and found out that the keySet size() operation is not expensive at all. So I will upload a new commit soon.